### PR TITLE
Improve Regional Partner search

### PR DESCRIPTION
**Pull request by clareconstantine**
We noticed that a surprising and unacceptable number of zipcodes for partners with statewide mappings in our db were not returning the correct results in this [search tool](https://code.org/educate/professional-learning/middle-high). This meant that teachers interested in our PD were not getting appropriately matched to partners, which might discourage them from applying.

For the failing zip codes, we were using the geocoder to tell us what state the zip code is in, which allows us to find the right regional partner. The geocoder returns multiple results, however, and in this case, we were using the first result without verifying that it was an American state, which is why we couldn't find the partners.

The fix: since this search is only for regional partners in the US, we need to make sure we are using the US result from the geocoder.

## Links

- [jira](https://codedotorg.atlassian.net/browse/PLC-1097?atlOrigin=eyJpIjoiZmZlODE2NTY0NzYyNGY0OTk3NzgzNjUwZjgyZDIwZWYiLCJwIjoiaiJ9)

## Testing story

I verified the fix by running the new line in the production console and verifying that a number of previously failing zip codes now get appropriately matched.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by molly-moen — APPROVED** [review]
Glad it was such a simple fix! Good find!

File: dashboard/app/models/regional_partner.rb [hunk 1]
```diff
@@ -194,7 +194,7 @@ def self.find_by_zip(zip_code_raw)
             # Geocoder can raise a number of errors including SocketError, with a common base of StandardError
             # See https://github.com/alexreisner/geocoder#error-handling
             Retryable.retryable(on: StandardError) do
-              state = Geocoder.search(zip_code)&.first&.state_code
+              state = Geocoder.search(zip_code).select {|result| result.country_code.downcase == "us"}&.first&.state_code
```
Comment (bencodeorg) [inline dashboard/app/models/regional_partner.rb]: Is country code always non-nil (and does a result always have a country code)?

**Comment by bencodeorg** [comment]
This seems like a good fix, and you may have already seen these, but wondering if @breville has any recollections around his similar work here? Any gotchas that we might run into with the change here?

https://github.com/code-dot-org/code-dot-org/pull/30898
https://github.com/code-dot-org/code-dot-org/pull/32855

**Comment by bencodeorg** [comment]
Also, out of curiosity, do we know if this is a new issue? Or did we just not notice before?

File: dashboard/app/models/regional_partner.rb (hunk 1 above)
Comment (clareconstantine) [reply dashboard/app/models/regional_partner.rb]: ooh good point, I will add checks!

**Comment by clareconstantine** [comment]
> Also, out of curiosity, do we know if this is a new issue? Or did we just not notice before?

I think this could have happened when we switched to Mapbox earlier this year, or it could have just been a change with how the geocoder lookup returns results - either way, it seems like we were getting lucky before, because we were relying on it returning the US results first which isn't a guarantee.

**Comment by bencodeorg** [comment]
Ooh interesting, I did not realize the Mapbox switch moved all of our geocoding (I thought it was just maps).

**Review by bencodeorg — APPROVED** [review]

**Comment by breville** [comment]
I'm also super curious as to whether this was always a possibility, or just became an option with the Mapbox switch.

**Pull request merged**